### PR TITLE
Introduce configTRACK_TASK_MEMORY_ALLOCATIONS to track dynamic allocations per task

### PR DESCRIPTION
**Pull request by RichardBarry**
Description
-----------

Introduce configTRACK_TASK_MEMORY_ALLOCATIONS.  When this is set to 1 and either heap_4 or heap_5 is used then the number of dynamic memory allocations, the number of dynamic memory frees, the total amount of heap memory allocated by a task but not yet freed (by any task), and the maximum amount of heap memory a task has ever had allocated at one time are all tracked on a task by task basis.  This additional information is returned in the TaskStatus_t structure populated by the vTaskGetInfo() and the uxTaskGetSystemState() API functions.

uxBasePriority is now only included in the TaskStatus_t structure if configUSE_MUTEXES is set to 1.  ulRunTimeCounter is now only included in the TaskStatus_t structure if configGENERATE_RUN_TIME_STATS is set to 1.

Update how the TCB members get initialised.  Previously each member was initialised separately.  Now, as there are new members, memset() the entire structure to 0 when the structure is allocated, then only explicitly initialise structure members that are not initialised to zero.

Test Steps
-----------
Add prvExerciseTaskHeapAllocationStats() to the MSVC Win32 and MPS2 QEMU IAR demos - upstreamed to the FreeRTOS/FreeRTOS repo in a separate PR.

Related Issue
-----------
The PR that adds prvExerciseTaskHeapAllocationStats() to the FreeRTOS/FreeRTOS repo.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by RichardBarry** [comment]
Spell check is failing because of variable names.  Is spell check meant to check the C code, or just the comments?

**Comment by mingyue86010** [comment]
> the



> Spell check is failing because of variable names. Is spell check meant to check the C code, or just the comments?

Hi @RichardBarry Just checked with team. It looks like the current spell checker checks any line of code with comment. So if a C code line contains a comment it will check the spell of entire line inclduing the code itself. We are look for some more generic spell checker to replace the current one and hopfully can resolve the problem.

File: include/task.h [hunk 1]
```diff
@@ -153,15 +153,28 @@ typedef struct xTASK_PARAMETERS
...
+    UBaseType_t uxCurrentPriority;               /* The priority at which the task was running (may be inherited) when the structure was populated. */
+    StackType_t * pxStackBase;                   /* Points to the lowest address of the task's stack area. */
+    configSTACK_DEPTH_TYPE usStackHighWaterMark; /* The minimum amount of stack space that has remained for the task since the task was created.  The closer this value is to zero the closer the task has come to overflowing its stack. */
+
+    #if ( configUSE_MUTEXES == 1 )
+        UBaseType_t uxBasePriority; /* The priority to which the task will return if the task's current priority has been inherited to avoid unbounded priority inversion when obtaining a mutex.  Only valid if configUSE_MUTEXES is defined as 1 in FreeRTOSConfig.h. */
+    #endif
+
+    #if ( configGENERATE_RUN_TIME_STATS == 1 )
+        uint32_t ulRunTimeCounter; /* The total run time allocated to the task so far, as defined by the run time stats clock.  See https://www.FreeRTOS.org/rtos-run-time-stats.html.  Only valid when configGENERATE_RUN_TIME_STATS is defined as 1 in FreeRTOSConfig.h. */
```
Comment (RichardBarry) [inline include/task.h:169]: ulRunTimeCounter should now have type configRUN_TIME_COUNTER_TYPE so users can use a 64-bit counter if they wish.

**Review by n9wxu — APPROVED** [review]

**Comment by Skptak** [comment]
I think this work has been done at this point?
It looks like we have [traceMALLOC](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/d43062ba7862f6c354ca414b411958ce12a7187f/portable/MemMang/heap_4.c#L326) and [traceFREE](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/d43062ba7862f6c354ca414b411958ce12a7187f/portable/MemMang/heap_4.c#L388) calls inside of the heap files now
Should this PR be closed?

**Comment by aggarg** [comment]
Closing this PR as it is possible to track allocations using trace macros - 

```c
#define NUM_ALLOCATION_ENTRIES          1024
#define NUM_PER_TASK_ALLOCATION_ENTRIES 32

/*
 * +-----------------+--------------+----------------+-------------------+
 * | Allocating Task | Entry in use | Allocated Size | Allocated Pointer |
 * +-----------------+--------------+----------------+-------------------+
 * |                 |              |                |                   |
 * +-----------------+--------------+----------------+-------------------+
 * |                 |              |                |                   |
 * +-----------------+--------------+----------------+-------------------+
 */
typedef struct AllocationEntry
{
    BaseType_t inUse;
    TaskHandle_t allocatingTask;
    size_t allocatedSize;
    void * allocatedPtr;
} AllocationEntry_t;

AllocationEntry_t allocationEntries[ NUM_ALLOCATION_ENTRIES ];

/*
 * +------+-----------------------+----------------------+
 * | Task | Memory Currently Held | Max Memory Ever Held |
 * +------+-----------------------+----------------------+
 * |      |                       |                      |
 * +------+-----------------------+----------------------+
 * |      |                       |                      |
 * +------+-----------------------+----------------------+
 */
typedef struct PerTaskAllocationEntry
{
    TaskHandle_t task;
    size_t memoryCurrentlyHeld;
    size_t maxMemoryEverHeld;
} PerTaskAllocationEntry_t;

PerTaskAllocationEntry_t perTaskAllocationEntries[ NUM_PER_TASK_ALLOCATION_ENTRIES ];

/*-----------------------------------------------------------*/

void TrackAllocation( size_t allocatedSize, void * ptr );
void TraceFree( void * ptr );

/*-----------------------------------------------------------*/

void TrackAllocation( size_t allocatedSize, void * ptr )
{
    size_t i;
    TaskHandle_t allocatingTask;
    AllocationEntry_t * pAllocationEntry = NULL;
    PerTaskAllocationEntry_t * pTaskAllocationEntry = NULL;

    if( xTaskGetSchedulerState() != taskSCHEDULER_NOT_STARTED )
    {
        allocatingTask = xTaskGetCurrentTaskHandle();

        for( i = 0; i < NUM_ALLOCATION_ENTRIES; i++ )
        {
            if( allocationEntries[ i ].inUse == pdFALSE )
            {
                pAllocationEntry = &( allocationEntries[ i ] );
                break;
            }
        }

        /* Do we already have an entry in the per task table? */
        for( i = 0; i < NUM_PER_TASK_ALLOCATION_ENTRIES; i++ )
        {
            if( perTaskAllocationEntries[ i ].task == allocatingTask )
            {
                pTaskAllocationEntry = &( perTaskAllocationEntries[ i ] );
                break;
            }
        }

        /* We do not have an entry in the per task table. Find an empty slot. */
        if( pTaskAllocationEntry == NULL )
        {
            for( i = 0; i < NUM_PER_TASK_ALLOCATION_ENTRIES; i++ )
            {
                if( perTaskAllocationEntries[ i ].task == NULL )
                {
                    pTaskAllocationEntry = &( perTaskAllocationEntries[ i ] );
                    break;
                }
            }
        }

        /* Ensure that we have space in both the tables. */
        configASSERT( pAllocationEntry != NULL );
        configASSERT( pTaskAllocationEntry != NULL );

        pAllocationEntry->allocatingTask = allocatingTask;
        pAllocationEntry->inUse = pdTRUE;
        pAllocationEntry->allocatedSize = allocatedSize;
        pAllocationEntry->allocatedPtr = ptr;

        pTaskAllocationEntry->task = allocatingTask;
        pTaskAllocationEntry->memoryCurrentlyHeld += allocatedSize;
        if( pTaskAllocationEntry->maxMemoryEverHeld < pTaskAllocationEntry->memoryCurrentlyHeld )
        {
            pTaskAllocationEntry->maxMemoryEverHeld = pTaskAllocationEntry->memoryCurrentlyHeld;
        }
    }
}

/*-----------------------------------------------------------*/

void TraceFree( void * ptr )
{
    size_t i;
    AllocationEntry_t * pAllocationEntry = NULL;
    PerTaskAllocationEntry_t * pTaskAllocationEntry = NULL;

    for( i = 0; i < NUM_ALLOCATION_ENTRIES; i++ )
    {
        if( ( allocationEntries[ i ].inUse == pdTRUE ) &&
            ( allocationEntries[ i ].allocatedPtr == ptr ) )
        {
            pAllocationEntry = &( allocationEntries [ i ] );
            break;
        }
    }

    /* Attempt to free a block that was never allocated. */
    configASSERT( pAllocationEntry != NULL );

    for( i = 0; i < NUM_PER_TASK_ALLOCATION_ENTRIES; i++ )
    {
        if( perTaskAllocationEntries[ i ].task == pAllocationEntry->allocatingTask )
        {
            pTaskAllocationEntry = &( perTaskAllocationEntries[ i ] );
            break;
        }
    }

    /* An entry must exist in the per task table. */
    configASSERT( pTaskAllocationEntry != NULL );

    pTaskAllocationEntry->memoryCurrentlyHeld -= pAllocationEntry->allocatedSize;

    pAllocationEntry->inUse = pdFALSE;
    pAllocationEntry->allocatingTask = NULL;
    pAllocationEntry->allocatedSize = 0;
    pAllocationEntry->allocatedPtr = NULL;
}

/*-----------------------------------------------------------*/
```

The following goes in FreeRTOSConfig.h:

```c
void TrackAllocation( size_t allocatedSize, void * ptr );
void TraceFree( void * ptr );

#define traceMALLOC( pvReturn, xAllocatedBlockSize ) \
TrackAllocation( xAllocatedBlockSize, pvReturn )

#define traceFREE( pv, xAllocatedBlockSize ) \
TraceFree( pv )
```

**Comment by Panometric** [comment]
@aggarg Will these macros be defined in a future release? Seems odd to close a PR with a workaround only documented in a closed PR. I think there is a strong case for having this functionality in the released version.

**Comment by aggarg** [comment]
@Panometric, the purpose of the above example code is to show the malloc tracking can be implemented in the application using trace macros. We will consider providing a sample with these definitions. I am interested to know your use case. Would you please be kind enough to share what you want to track and for what purpose?

**Comment by Panometric** [comment]
@aggarg Thanks, I do believe there is a general need since finding the uses of the heap can be quite daunting. In our case we are kind of forced to share the heap with a binary library for BLE that only provides hooks for malloc/free that are routed back through FreeRTOS. Allocated buffers are passed in that should be released by the library once sent.  It only leaks under certain conditions, like when comms were in progress when the connection drops.  So it's difficult problem to recreate. Therefore, we would use this do a post mortem, proving the the library vendor that they are causing the leak.

**Comment by aggarg** [comment]
Thank you for sharing your use case. Just one more question - do you keep this enabled in your production application or do you use it during development only?

**Comment by Panometric** [comment]
Hard to say for sure, probably only QA versions, since it introduces some load and memory usage.  But we would likely preserve the tracking data in for reporting after reset caused a failed malloc.

**Comment by aggarg** [comment]
Thank you for sharing.

**Pull request closed without merging**